traffic_ops/traffic_ops_golang/iso/iso.go

```diff
@@ -0,0 +1,368 @@
+		err := fmt.Errorf("invalid OS version directory: %q", ir.OSVersionDir)
+		api.HandleErr(w, req, tx.Tx, statusCode, err, err)
+		return
+	}
+
+	// Determine the kickstart root directory, which is either a default
+	// value or may be overridden by a database/Parameter entry.
+	ksDir, err := kickstarterDir(tx, ir.OSVersionDir)
+	if err != nil {
+		statusCode := http.StatusInternalServerError
+		userErr := errors.New("unable to determine kickstarter directory")
+		api.HandleErr(w, req, tx.Tx, statusCode, userErr, fmt.Errorf("%v: %v", userErr, err))
+		return
+	}
+
+	// cfgDir holds the kickstart config files within the root
+	// kickstart directory.
+	cfgDir := filepath.Join(ksDir, ksCfgDir)
+	log.Infof("cfg_dir: %s", cfgDir)
+
+	genISOCmd, err := newStreamISOCmd(ksDir)
+	if err != nil {
+		statusCode := http.StatusInternalServerError
+		userErr := errors.New("unable to initialize genISO command")
+		api.HandleErr(w, req, tx.Tx, statusCode, userErr, fmt.Errorf("%v: %v", userErr, err))
+		return
+	}
+	defer genISOCmd.cleanup()
+
+	// Allow for the request context to carry a modifier function that can change the
+	// genISOCmd's command. This is purely used for testing.
+	if cmdMod, ok := req.Context().Value(cmdOverwriteCtxKey).(func(in *exec.Cmd) *exec.Cmd); ok {
+		genISOCmd.cmd = cmdMod(genISOCmd.cmd)
+	}
+
+	log.Infof("Using %s ISO generation command: %s", genISOCmd.cmdType, genISOCmd.String())
+
+	if err = writeKSCfgs(cfgDir, ir, genISOCmd.String()); err != nil {
+		statusCode := http.StatusInternalServerError
+		userErr := errors.New("unable to create kickstarter files")
+		api.HandleErr(w, req, tx.Tx, statusCode, userErr, fmt.Errorf("%v: %v", userErr, err))
+		return
+	}
+
+	isoFilename := fmt.Sprintf("%s-%s.iso", ir.fqdn(), ir.OSVersionDir)
+	// strings.ReplaceAll was added in Go 1.12
+	isoFilename = strings.Replace(isoFilename, "/", "_", -1)
+
+	w.Header().Set(httpHeaderContentDisposition, fmt.Sprintf("attachment; filename=%q", isoFilename))
+	w.Header().Set(httpHeaderContentType, httpHeaderContentDownload)
+
+	if err = genISOCmd.stream(w); err != nil {
+		statusCode := http.StatusInternalServerError
+		userErr := errors.New("unable to generate ISO")
+		api.HandleErr(w, req, tx.Tx, statusCode, userErr, fmt.Errorf("%v: %v", userErr, err))
+		return
+	}
+
+	// Create changelog entry
+	err = api.CreateChangeLogBuildMsg(
+		api.ApiChange,
+		api.Created,
+		user,
+		tx.Tx,
+		"ISO",
+		ir.fqdn(),
+		map[string]interface{}{"OS": ir.OSVersionDir},
+	)
+	if err != nil {
+		// At this point, it's not possible to modify the HTTP response.
+		log.Errorf("error creating changelog entry for ISO creation: %v", err)
+	}
+}
+
+// writeRespErrorAlerts writes to w an Alerts JSON response with
+// an error level. Note: api.WriteRespAlertObj could not be used here
+// because it accepts a single "alert", whereas this response
+// may contain multiple alerts.
+func writeRespErrorAlerts(w http.ResponseWriter, req *http.Request, errMsgs []string) {
```

Comment:
I think this could be done using [`api.WriteAlerts`](https://godoc.org/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api#WriteAlertsObj).